(refactor) Update testCase to include pickle object

gd46 opened the pull request:
Resolves these issues: 

[Result is missing scenario & tags](https://github.com/cucumber/cucumber-js/issues/923)
[testCase does not contain Scenario name ](https://github.com/cucumber/cucumber-js/issues/943)

resolves #923

charlierudolph commented:
Can you please add a test or update the existing tests [here](https://github.com/cucumber/cucumber-js/blob/6c8bad20da4ca973805e23ade1d98c953d4c233a/features/hook_parameter.feature) to showcase the use of `pickle`?

gd46 commented:
@charlierudolph I updated the feature tests for hooks to include some assertions for pickle. I also updated the hooks readme to include some more info on what testCase contains, let me know what you think. 

Do we need to update anything in test_case_runner_spec.js  ? The first time I tried to include pickle in those tests as well the test was still failing. Not sure if I missed something on my original attempt.

manu7691 commented:
Hey @gd46 , any way to add also failure exception when status is failed
"result": {
    "duration": 155557,
    "status": "failed"
    "failureException": { name, stack}
  }

Thanks in advance,

charlierudolph commented:
@manu7691 that is already available in the after hook under `testCase.result` which has the structure: `{duration, exception, status}` where exception is only there when status is FAILED

manu7691 commented:
@charlierudolph I have been testing it with example for node, and it's still missing.
I noticed that */runtime/step_runner.js is returning {status: 'failed', exception, duration}: testStepResult, however on */runtime/test_case_runner.js only {status: 'failed', duration} is returned.

I think it would be better to return exception as well, to know the cause of failing test on AfterHooks.

![image](https://user-images.githubusercontent.com/1011682/31380650-3ab8d7d2-adb2-11e7-9681-8fc6047897f4.png)

Should I create an issue for this? Thanks

charlierudolph commented:
@manu7691 yes please create a separate issue for that

gd46 commented:
@charlierudolph Im going to tweak some more of the README soon. But is there anything we need to update. I wasnt sure what to do with the test_runner_spec. I noticed that had some tests around this as well, and my first try at updating them didnt work. And I was going to look at updating this readme as well: [api reference](https://github.com/cucumber/cucumber-js/blob/master/docs/support_files/api_reference.md)

charlierudolph commented:
I don't think you need to update the readme. You don't need to update test_runner_spec either. Updating the api reference is probably a good idea and you can mention that pickle comes from the [gherkin](https://github.com/cucumber/cucumber/tree/gherkin-v4.1.3/gherkin) library

gd46 commented:
Sounds good. @charlierudolph I removed the update I made to hooks readme. And I updated the api readme to include the pickle object in both the before and after hook. And I added a link in the After hooks docs to mention that pickle comes from gherkin library.

charlierudolph commented on docs/support_files/api_reference.md:
```diff
@@ -36,8 +36,9 @@ Defines a hook which is run after each scenario.
   * `timeout`: A hook-specific timeout, to override the default timeout.
 * `fn`: A function, defined as follows:
-  * The first argument will be an object of the form `{sourceLocation: {line, uri}, result: {duration, status}}` matching the event data for `test-case-finished`
+  * The first argument will be an object of the form `{sourceLocation: {line, uri}, result: {duration, status}, pickle: { tags: [{ name, location: { line, column } }], name, language, locations: [{ line, column }], steps: [{ text, arguments: [], locations: [{ line, column }] }]}}` matching the event data for `test-case-finished`
```
Please remove the text: "matching the event data for `test-case-finished`" as that is no longer true. Same for the Before Hook.

gd46 replied:
I removed that text. What is the event data for, and how come it no longer needs to match? I just want to understand what it is used for.

charlierudolph commented on docs/support_files/api_reference.md:
```diff
@@ -60,7 +61,7 @@ Multiple `AfterAll` hooks are executed in the **reverse** order that they are de
 #### `Before([options,] fn)`
 
-Defines a hook which is run before each scenario. Same interface as `After` except the first argument passed to `fn` will be an object of the form `{sourceLocation: {line, uri}}` matching the event data for `test-case-started`.
+Defines a hook which is run before each scenario. Same interface as `After` except the first argument passed to `fn` will be an object of the form `{sourceLocation: {line, uri}, pickle: { tags: [{ name, location: { line, column } }], name, language, locations: [{ line, column }], steps: [{ text, arguments: [], locations: [{ line, column }] }]}}` matching the event data for `test-case-started`.
```
Thoughts on saying the "will be an object of the form `{sourceLocation, pickle}` (see After hook for more details)" instead of repeating the pickle. Alternatively can reference the other way where they are defined only here and not in the After hook.

gd46 replied:
Agreed I cleaned it up. I put the details in the after hook since that is what comes first in the doc.

aslakhellesoy commented:
Hi @gd46,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

gd46 commented:
@aslakhellesoy thank you so much for the warm welcome, and allowing me to be apart of the cucumber community. I have been using this library for a long time and have watched it grow, and love where it is now. I hope I can help make further contributions going further.
